_posts/2021-11-23-how-to-dry-your-ref-arch.adoc

```diff
@@ -0,0 +1,588 @@
+file. For each child Terragrunt configuration:
+
+. Before making any changes, use the `render-json` command (`terragrunt render-json --terragrunt-json-out
+original.terragrunt.hcl.json`) to snapshot a copy of the current configuration with all the blocks and attributes
+rendered. The output of `render-json` is different from the one generated with `hcl2json` because it represents the
+effective Terragrunt configuration, with all expressions evaluated. We will use this to sanity check the refactored
+version.
+. Remove all the blocks, attributes, and input keys you identified in <<step_2>>.
+. For each remaining `locals`, check if the local variable is used in the configuration. If you find no references for
+the given local, remove it from the block.
+. Add an include block to import the common configuration for the component. To do this, you will want to use the
+relative path from the root Terragrunt configuration:
+
++
+[source,hcl]
+----
+include "envcommon" {
+  # Get to the root dir of the project by taking the directory of the root Terragrunt configuration found using
+  # find_in_parent_folders().
+  path = "${dirname(find_in_parent_folders())}_envcommon/networking/vpc.hcl"
+}
+----
+
+. Update the root `include` block with a label, if it doesn't have one. E.g., if you had:
+
++
+[source,hcl]
+----
+include {
+  path = find_in_parent_folders()
+}
+----
+
++
+Add the label `"root"` to the block:
+
++
+[source,hcl]
+----
+include "root" {
+  path = find_in_parent_folders()
+}
+----
+
+. Sanity check the resulting Terragrunt configuration by regenerating the json output (`terragrunt render-json
+--terragrunt-json-out updated.terragrunt.hcl.json`). This should be semantically equivalent to the original snapshot you
+created. Use https://github.com/josephburnett/jd[jd] to verify that the json files are semantically equivalent.
+
+. Run through a final sanity check of the resulting Terragrunt configuration by running `terragrunt validate` and
+`terragrunt plan`. There should be no differences resulting from configuration drift.
+
+At this point, your child Terragrunt configuration should be significantly smaller, with the bulk of the logic being
+moved to the common component configuration.
+
+
+[[further_steps]]
+==== (Optional) Further DRY configuration
+
+This guide takes a conservative approach to refactoring the Terragrunt configurations. However, there are additional
+Terragrunt features that can be leveraged to further DRY your configuration. Note that these features requires a bit
+more planning and thinking than a step by step tutorial can offer, and thus we can not provide a step by step tutorial
+as we offered above. Instead, we will describe the feature that Terragrunt offers with an example, and you can use that
+as a guide to identify further blocks and attributes that can be moved to the common component configuration.
+
+* <<deep_merge>>
+* <<expose_include>>
```

Comment:
NIT:

You can make this look like this in the UI:
![image](https://user-images.githubusercontent.com/13165182/144324710-f4872c9f-2c27-4dad-82f1-ae1c17de4339.png)

Reply:
This is done! This was automatically handled when the related heading was updated to use code snippet.